resource/wrapper/retryresource/crud_resource_test.go

```diff
@@ -1,364 +1,446 @@
-//	r.Order = append(r.Order, m)
-//
-//	p := framework.NewPatch()
-//	p.SetCreateChange("test create data")
-//	p.SetUpdateChange("test update data")
-//	p.SetDeleteChange("test delete data")
-//	return p, nil
-//}
-//
-//func (r *testCRUDResource) ApplyCreateChange(ctx context.Context, obj, createState interface{}) error {
-//	m := "ApplyCreatePatch"
-//	r.Order = append(r.Order, m)
-//
-//	if r.returnErrorFor(m) {
-//		return r.Error
-//	}
-//
-//	return nil
-//}
-//
-//func (r *testCRUDResource) ApplyDeleteChange(ctx context.Context, obj, deleteState interface{}) error {
-//	m := "ApplyDeletePatch"
-//	r.Order = append(r.Order, m)
-//
-//	if r.returnErrorFor(m) {
-//		return r.Error
-//	}
-//
-//	return nil
-//}
-//
-//func (r *testCRUDResource) ApplyUpdateChange(ctx context.Context, obj, updateState interface{}) error {
-//	m := "ApplyUpdatePatch"
-//	r.Order = append(r.Order, m)
-//
-//	if r.returnErrorFor(m) {
-//		return r.Error
-//	}
-//
-//	return nil
-//}
-//
-//func (r *testCRUDResource) returnErrorFor(errorMethod string) bool {
-//	ok := r.Error != nil && r.ErrorCount > r.errorCount && r.ErrorMethod == errorMethod
-//
-//	if ok {
-//		r.errorCount++
-//		return true
-//	}
-//
-//	return false
-//}
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"testing"
+
+	"github.com/giantswarm/backoff"
+	"github.com/giantswarm/micrologger/microloggertest"
+
+	"github.com/giantswarm/operatorkit/controller"
+	"github.com/giantswarm/operatorkit/resource"
+	"github.com/giantswarm/operatorkit/resource/crud"
+)
+
+// Test_Resource_Wrapper_RetryResource_ProcessDelete_ResourceOrder_RetryOnError
+// ensures the resource's methods are executed as expected when retrying the
+// deletion process.
+func Test_Resource_Wrapper_RetryResource_ProcessDelete_ResourceOrder_RetryOnError(t *testing.T) {
+	testCases := []struct {
+		ErrorCount          int
+		ErrorMethod         string
+		ExpectedMethodOrder []string
+	}{
+		{
+			ErrorCount:  1,
+			ErrorMethod: "GetCurrentState",
+			ExpectedMethodOrder: []string{
+				"GetCurrentState",
+				"GetCurrentState",
+				"GetDesiredState",
+				"NewDeletePatch",
+				"ApplyCreatePatch",
+				"ApplyDeletePatch",
+				"ApplyUpdatePatch",
+			},
+		},
+		{
+			ErrorCount:  2,
+			ErrorMethod: "GetCurrentState",
+			ExpectedMethodOrder: []string{
+				"GetCurrentState",
+				"GetCurrentState",
+				"GetCurrentState",
+				"GetDesiredState",
+				"NewDeletePatch",
+				"ApplyCreatePatch",
+				"ApplyDeletePatch",
+				"ApplyUpdatePatch",
+			},
+		},
+		{
+			ErrorCount:  2,
+			ErrorMethod: "ApplyDeletePatch",
+			ExpectedMethodOrder: []string{
+				"GetCurrentState",
+				"GetDesiredState",
+				"NewDeletePatch",
+				"ApplyCreatePatch",
+				"ApplyDeletePatch",
+				"GetCurrentState",
+				"GetDesiredState",
+				"NewDeletePatch",
+				"ApplyCreatePatch",
+				"ApplyDeletePatch",
+				"GetCurrentState",
+				"GetDesiredState",
+				"NewDeletePatch",
+				"ApplyCreatePatch",
+				"ApplyDeletePatch",
+				"ApplyUpdatePatch",
```

Comment:
This here is weird. The very old tests defined a different step execution. Did this change or did the refactoring of the wrapping break it? To me it looks like both ways could be argued for but I think what the test proves now is what we see in production and I am not sure at this point why the old tests defined it differently.

Reply:
Why is this weird? `ApplyDeletePatch` failed two times. Finally it succeeded and `ApplyUpdatePatch` got executed.

Reply:
What I mean with weird is that the assumption of the test changed. As I said, both ways can be argued for and I am kind of fine with either way as I do not see which one is better or worse.

Reply:
Sorted with Pawel's fix merged.